quickwit-indexing/src/source/logrotate_source.rs

```diff
@@ -0,0 +1,984 @@
+                Ok(true)
+            }
+            Either::Right(true) => Ok(false),
+            Either::Right(false) => {
+                anyhow::bail!(
+                    "find_next_file said it was fine to read {:?} after {}, but now it says it \
+                     isn't",
+                    self.latest_file,
+                    self.filename
+                );
+            }
+        }
+    }
+
+    /// verify if we are reading the file we should. Returns true if we are.
+    /// If !is_next, it's necessarily the right file. Else it's the right file only if no
+    /// other file appeared that we should read. This function must be called after opening
+    /// self.latest_file in case log got rotated between the decision of resolve_next and
+    /// the call to File::open.
+    async fn revalidate_next_file(&self) -> Result<bool, anyhow::Error> {
+        if !self.is_next {
+            // this is definitely the right file
+            return Ok(true);
+        }
+        let next_file = find_next_file(
+            &self.archived_dir,
+            &self.filename,
+            &self.name_pattern,
+            &self.latest_file,
+        )
+        .await?;
+        let ok = matches!(next_file, Either::Right(true));
+        Ok(ok)
+    }
+
+    /// The path of the file to read
+    fn filepath(&self) -> PathBuf {
+        if self.is_next {
+            self.latest_file.clone()
+        } else {
+            let mut res = self.archived_dir.clone();
+            res.push(&self.filename);
+            res
+        }
+    }
+}
+
+/// Try to find the next file to read from.
+/// Returns Ok(Left(filename)) if a filename was found, Ok(Right(true)) if no file was found,
+/// and it's safe to read from latest_log_file, or Ok(Right(false)) when no file was found and
+/// reading from latest_log_file could cause skipping some files.
+async fn find_next_file(
+    dir: &Path,
+    current_file: &str,
+    pattern: &GlobPattern,
+    latest_log_file: &Path,
+) -> Result<Either<String, bool>, io::Error> {
+    // first check len of latest log file. Then read data about all file which could interest us.
+    let len = fs::metadata(latest_log_file)
+        .await
+        .map(|f| f.len())
+        .unwrap_or(0);
+    let mut list_dir = fs::read_dir(dir).await?;
+    let mut files = BTreeMap::new();
+    let mut found_current_file = false;
+    let mut found_any_log = false;
+    while let Some(entry) = list_dir.next_entry().await? {
+        let filename = if let Ok(name) = entry.file_name().into_string() {
+            name
+        } else {
+            continue;
+        };
+        if filename == current_file {
+            found_current_file = true;
+        }
+        if pattern.matches(&filename) {
+            found_any_log = true;
+            if filename.as_str() > current_file {
+                // this is a log file which is newer than current file
+                let len = entry.metadata().await?.len();
```

Comment:
If I understand correctly, we identify the rotation by globlisting the directory, and picking the first file (lexicographically) that is above our last processed files?

Was inspecting `creation_time` an option too?

Can we add explanation about that somewhere?

Reply:
I'm generally reluctant to use filesystem time, I find it often leads to broken assumptions. As a case that would definitely break (but is not supported yet anyway) when logrotate compresses older log files, it creates a new files, reading `creation_time` could cause an old log file getting compressed to be considered as newer than a recent (uncompressed) log file.
There would also be issue with what the checkpoints should then look like (as it would relax the requirement on lexicographic ordering of file names). Replacing the filename with creation time seems rather fragile

Reply:
> reading creation_time could cause an old log file getting compressed to be considered as newer than a recent (uncompressed) log file.

That's not what I meant.
My point is that we could us `creation_time` to identify that the current file has been moved to another file.

```
for file in list_dir() {
   if file.creation_time == current_tail_file_creation_time {
      // we rotated
   }
}
```

I agree all solutions are muddy though. Let's stick to yours!